hw/ip/clkmgr/data/clkmgr.sv.tpl

```diff
@@ -0,0 +1,179 @@
+// Copyright lowRISC contributors.
+// Licensed under the Apache License, Version 2.0, see LICENSE for details.
+// SPDX-License-Identifier: Apache-2.0
+//
+// The overall clock manager
+
+`include "prim_assert.sv"
+
+<%
+clks_attr = cfg['clocks']
+srcs = clks_attr['srcs']
+grps = clks_attr['groups']
+
+## feedthrough clocks
+ft_clks   = {clk:src for grp in grps for (clk,src) in grp['clocks'].items()
+  if grp['name'] == 'powerup'}
+
+## root-gate clocks
+rg_clks   = {clk:src for grp in grps for (clk,src) in grp['clocks'].items()
+  if grp['name'] != 'powerup' and grp['sw_cg'] == 'no'}
+
+## direct sw control clocks
+sw_clks   = {clk:src for grp in grps for (clk,src) in grp['clocks'].items()
+  if grp['sw_cg'] == 'yes'}
+
+## sw hint clocks
+hint_clks = {clk:src for grp in grps for (clk,src) in grp['clocks'].items()
+  if grp['sw_cg'] == 'hint'}
+
+%>
+
+module clkmgr import clkmgr_pkg::*; (
+  // Primary module clocks
+  input clk_i,
+  input rst_ni,
+% for src in srcs:
+  input clk_${src['name']}_i,
+% endfor
+
+  // Bus Interface
+  input tlul_pkg::tl_h2d_t tl_i,
+  output tlul_pkg::tl_d2h_t tl_o,
+
+  // pwrmgr interface
+  input pwrmgr_pkg::pwr_clk_req_t pwr_i,
+  output pwrmgr_pkg::pwr_clk_rsp_t pwr_o,
+
+  // dft interface
+  input clk_dft_t dft_i,
+
+  // idle hints
+% for k,v in hint_clks.items():
+  input ${k}_status_t ${k}_status_i,
+% endfor
+
+  // clock output interface
+  output clk_out_req_t clocks_o
+
+);
+
+  ////////////////////////////////////////////////////
+  // Register Interface
+  ////////////////////////////////////////////////////
+
+  clkmgr_reg_pkg::clkmgr_reg2hw_t reg2hw;
+  clkmgr_reg_pkg::clkmgr_hw2reg_t hw2reg;
+
+  clkmgr_reg_top i_reg (
+    .clk_i,
+    .rst_ni,
+    .tl_i,
+    .tl_o,
+    .reg2hw,
+    .hw2reg,
+    .devmode_i(1'b1)
+  );
+
+  ////////////////////////////////////////////////////
+  // Power up group - feedthrough
+  ////////////////////////////////////////////////////
+  ## This group is hardcoded right now because it is the only one.
+  ## In the future if there are more, feedthrough should be converted to an attribute
+
+% for k,v in ft_clks.items():
+  assign clocks_o.${k} = clk_${v}_i;
+% endfor
+
+  ////////////////////////////////////////////////////
+  // Root gating
+  ////////////////////////////////////////////////////
+
+  // the rst_ni connection below is incorrect, need to find a proper reset in the sequence to use
```

Comment:
this reset thing really sucks...
I started off assuming clkmgr is NOT always on. But this means it needs a non-always-on reset. However the timing of this non-always-on reset needs to be ahead of lc/sys, otherwise clocks won't turn on.

So for the first draft i may make this always on...that might be easier also since i handle aon clks in the same place

Reply:
SG